### ISSUE TITLE
Built-in support for Transform

### ISSUE THREAD

**Issue by alt-romes** [alt-romes/ghengin#4]
We should allow `Transform` to be used as a property binding value, just like we allow `Mat4`.
This likely means using the new type-class which is not Syntactic (see #3) and making `Transform` compatible with `M 4 4` in the Compatible families.

I wonder if it would also be possible for user-defined types to extend the compatible type family -- even though Transform should be built in.

**Comment by walseb** [comment]
Do you mean something like this?

```
instance ShaderData Transform where
  type FirType Transform = M 4 4 Float
```
For me it automatically converts all `Transforms` into `M 4 4 Float` when sent to the shader.

**Comment by alt-romes** [comment]
Yes, something like that.

I think my worry is that you can't receive `M 4 4 Float` directly on the shader. You need a struct with a field `[ fld :-> M 4 4 Float]`.

Can you still use it like this how you've defined it?

**Comment by walseb** [comment]
Yeah, I think FIR itself needs that `M 4 4 Float` to be wrapped in a field, I think that's unavoidable. I got an error before from FIR when I tried removing the field/struct part from the shader description.

I have this in my FIR shader description:
```haskell
'[ "model" ':-> Uniform '[DescriptorSet 2, Binding 0 ] (Struct '[ "m" ':-> M 4 4 Float ]) ]
```

I can then push it like this from Ghengin with createMesh using the instance above:
```haskell
G.StaticBinding (Ur (InStruct @"m" (scale 1))
```

**Comment by alt-romes** [comment]
Ah! That sounds good, using InStruct + M 4 4 Float for Transform

**Comment by walseb** [comment]
I realize now I misread what you said before. 

About having `Transform` expand into `[ fld :-> M 4 4 Float]` automatically, without the need for `InStruct`. I tried that before, and it makes some types very annoying to work with. For instance, if you declare this type in a FIR shader input: 
```haskell
(Struct '[ "m" ':-> Array 50 (M 4 4 Float)])
```

You'd think this `DynamicBind` on the Ghengin end would work to match the type: 
```haskell
Gsd.InStruct "m" (Array 50 Transform)

```
But if `Transform` automatically adds a field, you will actually end up with this on ShaderData expansion:
```haskell
Gsd.InStruct "m" (Array 50 [ fld :-> M 4 4 Float])
```
Which is an unecessary layer..

But if ShaderData never adds fields automatically, you get this expansion instead:
```haskell
Gsd.InStruct "m" (Array 50 (M 4 4 Float))
```

So I think while adding the field automatically would make simple types easier, it makes some nested types impossible. But perhaps I'm not seeing some better solution here.

BTW, my use case for that exact type is to store the position of 50 objects, which I then index into in my instance shader.

**Comment by alt-romes** [comment]
Perfect! Let's do your suggestion of `M 4 4 Float` directly! Happy to receive a patch for that.

**Comment by alt-romes** [comment]
Closed by #41
